Teach cachetasks to update requests if needed

marcoscaceres opened the pull request:
- Brought back CacheTask.update() with a vengeance :feelsgood:.

marcoscaceres committed:
Teach cachetasks to update requests if needed

marcoscaceres commented:
@oyiptong r?

oyiptong commented on src/js/lib/cachetasks.js:
```diff
@@ -191,6 +191,27 @@
+        let response = yield cache.match(request);
+        let shouldUpdate = !response || options.force || this.isStale(response);
+        if (shouldUpdate) {
+          response = yield fetch(request);
```
Should the fetch fail, what should be the behavior?
Respond with an error and blow the cache?

In this case, I think the prudent option would be to return the cached request even if it's stale, to ensure a working application.

This is a scenario which needs to be defined and tested.

marcoscaceres replied:
> In this case, I think the prudent option would be to return the cached request even if it's stale, to ensure a working application.

Yeah, I thought of doing that. My rationale was to allow the fetch to throw and allow the calling code to deal with it. Will add the recovery code into the method. 

> This is a scenario which needs to be defined and tested.

It's going to be hard to test, because we need to somehow force the Karma's test server to somehow produce a connection error :( Alternatively, we would need to make Gecko go-offline, but that would also break Karma as it relies on a socket connection :crying_cat_face: :sob:

marcoscaceres replied:
- Will add (fixed comment above)

marcoscaceres commented:
@oyiptong, see updated code...

marcoscaceres commented:
Found a way to test it :dancer: :dancers: using a random port number/... I'm using "http://localhost:9999/". I'm open to using a alternative port number tho.

marcoscaceres commented:
@oyiptong updated tests to include error recovery scenarios + scenarios where it's not possible to recover (hence throws).

marcoscaceres commented:
fixed a couple of little bugs, should hopefully be good now... @oyiptong r?

oyiptong commented:
Maybe we should take a look at what other read-through caches like Cloudflare/Cloudfront/Squid/Varnish do, except also with an eye to keep things "always online".

Perhaps this is what an application would expect to remain offline:
- 200: Cache: updated
- 301, 308: we do the redirect (default fetch behavior?) and update the cache, both for the original request and the updated request. Cache: updated (both)
- 302, 307: we do the redirect and update the cache, only for the redirect request. Cache: updated (redirect only)
- 400, 401, 403, 405, 413 (basically, all errors): we return the errored response. Cache: untouched
- 404, 410: we invalidate the cache and return the errored response. The page must've moved or been deleted. We don't store anything in the cache. Cache: invalidated
- 500: we return the last good response. Cache: untouched

The idea is that the application needs to keep functioning even in the case that:
1. the network is down
2. the server is down

However, if there are any errors cause from the client, we should not cache (which is the current behavior)

## Cloudflare

On [301/302 on cloudflare](https://support.cloudflare.com/hc/en-us/articles/200168326-Are-301-and-302-redirects-cached-by-CloudFlare-):

> If no cache headers are provided (no Cache-Control or Expires) and the url is cacheable (.jpg, .css, .js etc.) then CloudFlare caches both 301 and 302s. We cache 301 for a couple of hours and 302s for a shorter period of time (~20 minutes).
> 
> However, if the headers are set, they override these defaults. In particular, Cache-Control: private, no-cache or public, max-age=0 (or some combination) will tell us not to cache.
> 
> For urls that do not get cached, if you want to cache the page, you can create a Page Rule and select 'cache everything'. This will make the resource cacheable, as long as the headers allow it.

They also have an [always online](https://www.cloudflare.com/always-online/) feature.

Here's an excerpt from [the page on how to trigger it](https://support.cloudflare.com/hc/en-us/articles/202238800):

> In order to trigger Always Online, your web server will need to be returning a standard HTTP Error code of 502 or 504 timeout. Always Online will also work when we encounter issues contacting your origin (CloudFlare Errors 521 & 523), timeouts (522 & 524), SSL errors (525 & 526) or an unknown error (520). Always Online will not be triggered for other HTTP response codes, such as 500, 503, database connection errors, internal server error, or empty replies from server.

Thoughts? (Also, let me know if the drugs I'm taking are too strong)

marcoscaceres commented:
Great suggestions! Comments below.

> On Nov 15, 2015, at 2:12 AM, Olivier Yiptong notifications@github.com wrote:
> 
> Maybe we should take a look at what other read-through caches like Cloudflare/Cloudfront/Squid/Varnish do, except also with an eye to keep things "always online".
> 
> Perhaps this is what an application would expect to remain offline:
> 
> 200: Cache: updated
> 301, 308: we do the redirect (default fetch behavior?) and update the cache, both for the original request and the updated request. Cache: updated (both)

Already handled by fetch.

> 302, 307: we do the redirect and update the cache, only for the redirect request. Cache: updated (redirect only)
> Handled automatically by fetch.
> 
> 400, 401, 403, 405, 413 (basically, all errors): we return the errored response. Cache: untouched
> Sounds ok to me, but if we ever hit these, it means we've screwed something up (I.e., these are effectively dev errors we would only encounter during development... Fingers-crossed-knock-on-wood😅)
> 
> 404, 410: we invalidate the cache and return the errored response. The page must've moved or been deleted. We don't store anything in the cache. Cache: invalidated
> As above. Returning the last good response would be good here, but naturally erroring all over the console (or possibly pinging an API end point to let us know things are broken... There is a spec in the works that does this automatically, but would be cool to add SW support for it).
> 
> 500: we return the last good response. Cache: untouched
> Agree, will add this.
> 
> The idea is that the application needs to keep functioning even in the case that:
> 1. the network is down
> 2. the server is down
1. We f'ed up (400).
2. CDN/network f'ed up (500).

1 and 2 already handled tho. But I like the idea of doing 3 and 4 for robustness 💪. So good idea!

> However, if there are any errors cause from the client, we should not cache (which is the current behavior)
> 
> Agree.
> 
> Cloudflare
> 
> On 301/302 on cloudflare:
> 
> If no cache headers are provided (no Cache-Control or Expires) and the url is cacheable (.jpg, .css, .js etc.) then CloudFlare caches both 301 and 302s. We cache 301 for a couple of hours and 302s for a shorter period of time (~20 minutes).
> 
> However, if the headers are set, they override these defaults. In particular, Cache-Control: private, no-cache or public, max-age=0 (or some combination) will tell us not to cache.
> 
> Could add more of the above (already do simple staleness checks on various headers), but this seems like overkill right now as we presumedly control what headers the CDN sends back. 

I'd be keen to add support for the bits we need and enhance the cache validation code depending on what we need. I don't want to end up rewriting the browsers HTTP cache in SW or basically we are doing something seriously wrong (or SWs are pointless, let's just use the browser's cache... Know what I mean?)

> For urls that do not get cached, if you want to cache the page, you can create a Page Rule and select 'cache everything'. This will make the resource cacheable, as long as the headers allow it.

Yeah, this:) 

> They also have an always online feature.
> 
> Here's an excerpt from the page on how to trigger it:
> 
> In order to trigger Always Online, your web server will need to be returning a standard HTTP Error code of 502 or 504 timeout. Always Online will also work when we encounter issues contacting your origin (CloudFlare Errors 521 & 523), timeouts (522 & 524), SSL errors (525 & 526) or an unknown error (520). Always Online will not be triggered for other HTTP response codes, such as 500, 503, database connection errors, internal server error, or empty replies from server.

IIUC, this basically means getting a 200 OK from the CDN when things f'up on the server... I imagine we would also enable this on our CDN, right?

> Thoughts?
> 
> Mostly sounds great - though a little more robust than I would have gone for (but certainly won't hurt and could save us down the line).  All bash something tomorrow. Thanks for the detailed thought-through on the design. 
> 
> —
> Reply to this email directly or view it on GitHub.

marcoscaceres committed:
Add cors and karma-express dev deps

marcoscaceres committed:
After registering, unregister, as to not screw up other tests

marcoscaceres committed:
Add express HTTP Server to karma + cors middleware

marcoscaceres committed:
Only cache responses that are OK when updating

marcoscaceres committed:
Fixup of src/js/lib/cachetasks.js

marcoscaceres committed:
Test HTTP status codes + add addAll error case

marcoscaceres committed:
Fix style issue, make codeclimate happy

marcoscaceres committed:
Fix jscs style issue with karmaconf

marcoscaceres commented:
@oyiptong, added handling for http errors. Basically: if response is "ok", then update... if it's not ok, then return the cached one (if any). If no cached response, throw.

oyiptong commented:
r+
